Documentation/building/dev-workflow.md

```diff
@@ -0,0 +1,46 @@
+Dev Workflow
+===============
+
+The dev workflow describes the development process to follow when working in our repo. 
+It is divided into specific tasks that are fast, transparent and easy to understand.
+
+This provides flexibility as with discrete tasks people can iterate on a specific one without needing to run the full workflow, other processes can reuse the tasks as needed and anyone can put together the tasks as wanted to create a customized workflow.
+
+## Process 
+![Dev Workflow process](../images/dev-workflow.png)
+
+All the tasks (except enlist, code, submit) will have scripts located at the root of the repo, and will have usage help to better understand what they provide.
+In addition, we want to have an “easy button” script for running many or all of these tasks sequentially in a single command.
+
+## Tasks
+
+**Clean**
+
+Cleans the local enlistment/environment, for example removing files not in the repository. 
+
+**Sync**
+
+The sync task gets the latest source history and packages in order to eliminate all the network traffic when we build. This way we are able to hit the network only when we are intentional about doing it and then be able to build even in offline mode. 
```

Comment:
Even when people have an internet connection all the time we have found that doing restore during the build introduces reliability issues and we are trying to improve our overall build reliability by preventing it from hitting the network during most build operations. With that said I do think there are some things we can do to make the current things better. For example when building an individual project we should actually check the timestamp of the project.json vs the project.lock.json file and do a local restore for that project if the project.json file is newer. This will enable devs to make local changes and build incrementally and only ever hit the network when it is actually needed and not on every build. 

We also know that doing root level sync/restores increases the overall performance of the build because we can batch the requests, so I still think that is something we should generally do at the root level and not on each individual project when building from the root. 

So then it just becomes a question of what we do by default. @maririos mentions an "easy mode" button which will end up being a single script people can run from the root, similar to what build.cmd does today, but we definitely need to break apart the different discreet operations, sync, build, test, etc into individual commands so that we can run them independently both locally, if we choose, and in our official builds. The easy script will just call the different individual building block scripts. 

We still have some things to work through but I don't think in the end we will negatively impact dev productivity. In fact the dev workflow changes are intended to improve the dev productivity by giving devs an easy repeatable set of steps to get their jobs done, so if they do negatively impact the dev productivity then we not meet our goals and we should make adjustments.